### PR TITLE
make a category of Dedekind domains, remove code from ring.pyx

### PR DESCRIPTION
**Pull request by fchapoton**
This is creating a `DedekindDomains` category and moving the associated code from `ring.pyx` to the category.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by fchapoton** [comment]
maybe one should make ZZ belong to this category ?

File: src/sage/categories/dedekind_domains.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                sage: K.<i> = QuadraticField(-1)
+                sage: R = K.order(2*i); R
+                Order of conductor 2 generated by 2*i
+                 in Number Field in i with defining polynomial x^2 + 1 with i = 1*I
+                sage: R.is_maximal()
+                False
+                sage: R.krull_dimension()
+                1
+            """
+            return 1
```
Comment (tscrim) [inline src/sage/categories/dedekind_domains.py]:
```suggestion
            return ZZ.one()
```
To avoid Python `int`s getting out.

**Review by tscrim — CHANGES_REQUESTED** [review]
One tiny thing (it should be fixed more broadly though). However, I think it is a good thing to refine the category of `ZZ` and remove duplicate methods (these should never be a critical part of any algorithm IMO, so downgrading Cython -> Python shouldn't have any real effect...I hope).

**Comment by fchapoton** [comment]
It seems that ZZ prefers to inherit `krull_dimension` from `rings/ring.pyx`. Same for `is_integrally_closed`. Not sure how to fix that.

**Comment by tscrim** [comment]
Ah, right, categories have the lowest precedence. Yes, we can revert removing them. It was just trying to avoid some redundancies.

**Review by tscrim — CHANGES_REQUESTED** [review]
Thank you.

One last thing, but unfortunately it is somewhat annoying. We have an import of `DedekindDomain` in `rings/all.py`, which makes its way into the global namespace. So we need to properly formally deprecate the class. *sigh*

**Review by tscrim — APPROVED** [review]
Thank you. Once the bot shows tests pass, you can set a positive review.

**Pull request merged**